crates/sui-indexer-alt-graphql/src/api/types/available_range.rs

```diff
@@ -0,0 +1,174 @@
+    }
+}
+
+impl AvailableRange {
+    /// Get retention information for a specific query type and field
+    pub(crate) fn new(
+        ctx: &Context<'_>,
+        scope: &Scope,
+        retention_key: RetentionKey,
+    ) -> Result<Self, RpcError> {
+        let watermarks: &Arc<Watermarks> = ctx.data()?;
+        let pipelines = pipeline(
+            &retention_key.type_,
+            &retention_key.field,
+            retention_key.filter.as_deref(),
+        );
+
+        let lo_checkpoint =
+            pipelines
+                .iter()
+                .try_fold(0, |acc: u64, pipeline| -> Result<u64, RpcError> {
+                    let watermark = watermarks.pipeline_lo_watermark(pipeline)?;
+                    let checkpoint = watermark.checkpoint();
+                    Ok(acc.max(checkpoint))
+                })?;
+
+        Ok(Self {
+            scope: scope.clone(),
+            first: lo_checkpoint,
+        })
+    }
+}
+
+/// Maps GraphQL query components to watermark pipeline names.
+///
+/// Determines which watermark pipelines are relevant for a given GraphQL query.
+/// The pipeline names are used to query watermark data to determine the
+/// checkpoint sequence range (available range) for which data is available.
+///
+fn pipeline(type_: &str, field: &str, filter: Option<&str>) -> &'static [&'static str] {
+    match (type_, field, filter) {
+        // Transaction queries
+        ("Query", "transaction", None) => &["tx_digests"],
+        ("Query", "transactions", None) => &["tx_digests"],
+        ("Query", "transactions", Some("affectedAddress")) => {
+            &["tx_digests", "tx_affected_addresses"]
+        }
+        ("Query", "transactions", Some("affectedObject")) => &["tx_digests", "tx_affected_objects"],
+        ("Query", "transactions", Some("function")) => &["tx_digests", "tx_calls"],
+        ("Query", "transactions", Some("kind")) => &[
+            "tx_digests",
+            "tx_kinds",
+            "tx_affected_addresses",
+            "tx_affected_objects",
+        ],
+        ("Query", "transactions", Some("sentAddress")) => &[
+            "tx_digests",
+            "tx_affected_addresses",
+            "tx_affected_objects",
+            "tx_calls",
+        ],
+        (_, "transactions", f) => pipeline("Query", "transactions", f),
+
+        // Checkpoint queries
+        ("Query", "checkpoints", _) => &["cp_sequence_numbers"],
+        ("Query", "checkpoint", None) => pipeline("Query", "checkpoints", None),
+        (_, "checkpoints", f) => pipeline("Query", "checkpoints", f),
+
+        // Event queries
+        ("Query", "events", None) => &["ev_struct_inst", "ev_emit_mod"],
+        ("Query", "events", Some("module")) => &["ev_emit_mod"],
+        ("Query", "events", Some("type")) => &["ev_emit_mod"],
+        ("Query", "events", Some("sender")) => pipeline("Query", "events", None),
+        (_, "events", f) => pipeline("Query", "events", f),
```

Comment:
Avoid catch alls, not the right approach.
`("Transaction", "events", f) => pipeline("Query", "events", f)` fails